### ISSUE TITLE
[Comprehensive Gap Analysis] Traditional ML Models - 60+ Missing Models Identified

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#319]
### User Story

> As a data scientist, I want access to a comprehensive suite of traditional machine learning models, including classification, regression, clustering, and dimensionality reduction algorithms, so that I can perform a wide range of data analysis tasks directly within AiDotNet.

---

## COMPREHENSIVE GAP ANALYSIS (February 2026)

This gap analysis compares AiDotNet's current Traditional ML implementations against scikit-learn 1.8, XGBoost, LightGBM, CatBoost, and recent research.

---

### Classification (`src/Classification/`)
| Subdirectory | Implemented Models | Status |
|--------------|-------------------|--------|
| `Trees/` | DecisionTreeClassifier | Complete |
| `Ensemble/` | RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier, ExtraTreesClassifier | Complete |
| `SVM/` | SupportVectorClassifier, NuSupportVectorClassifier, LinearSupportVectorClassifier | Complete |
| `NaiveBayes/` | GaussianNaiveBayes, MultinomialNaiveBayes, BernoulliNaiveBayes, CategoricalNaiveBayes, ComplementNaiveBayes | Complete |
| `Neighbors/` | KNeighborsClassifier | Complete |
| `DiscriminantAnalysis/` | LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis | Complete |
| `Meta/` | VotingClassifier, StackingClassifier, BaggingClassifier, OneVsOneClassifier, OneVsRestClassifier, ClassifierChain, MultiOutputClassifier | Complete |
| `Linear/` | SGDClassifier, PerceptronClassifier, RidgeClassifier, PassiveAggressiveClassifier | Complete |

### Regression (`src/Regression/`)
RidgeRegression, LassoRegression, ElasticNetRegression, BayesianRegression, DecisionTreeRegression, RandomForestRegression, GradientBoostingRegression, SupportVectorRegression, KNearestNeighborsRegression, PoissonRegression, NegativeBinomialRegression, QuantileRegression, IsotonicRegression, GaussianProcessRegression, etc. - **Complete**

### Clustering (`src/Clustering/`)
90+ models across Partitioning, Density, Hierarchical, Spectral, SemiSupervised, Streaming, Probabilistic - **Complete**

### Dimensionality Reduction (`src/Preprocessing/DimensionalityReduction/`)
35+ methods (PCA variants, TSNE, UMAP, PaCMAP, etc.) - **Complete**

### Anomaly Detection (`src/AnomalyDetection/`)
65+ detectors - **Complete**

### Interpretability (`src/Interpretability/`)
50+ explainers (SHAP, LIME, GradCAM, etc.) - **Complete**

---

## EXISTING INTERFACE PATTERNS (MUST FOLLOW)

All new interfaces **MUST inherit from `IFullModel<T, TInput, TOutput>`** either directly or through an existing interface.

**Current Pattern:**
```csharp
// Classification - extends IFullModel
public interface IClassifier<T> : IFullModel<T, Matrix<T>, Vector<T>> { ... }

// Multi-label - extends IClassifier (which extends IFullModel)
public interface IMultiLabelClassifier<T> : IClassifier<T> { ... }

// Regression - extends IFullModel
public interface IRegression<T> : IFullModel<T, Matrix<T>, Vector<T>> { ... }
```

---

### 1. Semi-Supervised Classification (HIGH PRIORITY)

**New Interface Required:**
```csharp
// src/Interfaces/ISemiSupervisedClassifier.cs
public interface ISemiSupervisedClassifier<T> : IClassifier<T>
{
    void TrainSemiSupervised(Matrix<T> labeledX, Vector<T> labeledY, Matrix<T> unlabeledX);
    int NumLabeledSamples { get; }
    int NumUnlabeledSamples { get; }
}
```

**Models:**
| Model | Description | Priority |
|-------|-------------|----------|
| SelfTrainingClassifier | Iteratively labels unlabeled data | High |
| LabelPropagation | Graph-based propagation | High |
| LabelSpreading | Soft-label propagation | High |

**Files to Create:**
```
src/Classification/SemiSupervised/
├── SemiSupervisedClassifierBase.cs  (extends ClassifierBase, implements ISemiSupervisedClassifier)
├── SelfTrainingClassifier.cs
├── LabelPropagation.cs
└── LabelSpreading.cs
```

---

### 2. Multi-Label Classification (HIGH PRIORITY)

**Interface Already Exists:** `src/Interfaces/IMultiLabelClassifier.cs` (extends `IClassifier<T>`)

**Models Needed:**
| Model | Description | Priority |
|-------|-------------|----------|
| BinaryRelevance | Independent binary classifiers | High |
| LabelPowerset | Each label combination as single class | High |
| RAkEL | Random k-labelsets | Medium |
| MLkNN | Multi-label k-NN | Medium |

**Files to Create:**
```
src/Classification/MultiLabel/
├── MultiLabelClassifierBase.cs  (extends ClassifierBase, implements IMultiLabelClassifier)
├── BinaryRelevance.cs
├── LabelPowerset.cs
├── RAkEL.cs
└── MLkNN.cs
```

---

### 3. Imbalanced Learning (HIGH PRIORITY)

**No new interface needed** - these are data augmentation/resampling techniques.

**Files to Create:**
```
src/Augmentation/Tabular/
├── AdasynAugmenter.cs
├── BorderlineSmoteAugmenter.cs
├── SvmSmoteAugmenter.cs
├── TomekLinksAugmenter.cs
├── SmoteTomekAugmenter.cs
├── SmoteEnnAugmenter.cs

src/Augmentation/Tabular/Undersampling/
├── RandomUnderSampler.cs
├── NearMissUnderSampler.cs

src/Classification/Ensemble/
├── BalancedRandomForestClassifier.cs  (extends EnsembleClassifierBase)
├── BalancedBaggingClassifier.cs
├── EasyEnsembleClassifier.cs
```

---

### 4. Survival Analysis (HIGH PRIORITY - Completely Missing)

**New Interfaces Required:**
```csharp
// src/Interfaces/ISurvivalModel.cs
// Note: Survival models predict time-to-event, different from standard regression
public interface ISurvivalModel<T> : IFullModel<T, Matrix<T>, Vector<T>>
{
    (Vector<T> survivalTimes, Vector<T> eventIndicators) PredictSurvival(Matrix<T> input);
    Vector<T> PredictHazard(Matrix<T> input);
    Vector<T> PredictCumulativeHazard(Matrix<T> input);
    T PredictMedianSurvivalTime(Vector<T> input);
}

// src/Interfaces/ISurvivalEstimator.cs
public interface ISurvivalEstimator<T> : IFullModel<T, (Vector<T> times, Vector<T> events), Vector<T>>
{
    Vector<T> SurvivalFunction(Vector<T> times);
    Vector<T> CumulativeHazardFunction(Vector<T> times);
    T MedianSurvivalTime { get; }
}
```

**Directory Structure (NEW):**
```
src/SurvivalAnalysis/
├── SurvivalModelBase.cs
├── Estimators/
│   ├── KaplanMeierEstimator.cs
│   └── NelsonAalenEstimator.cs
├── Regression/
│   ├── CoxProportionalHazards.cs
│   ├── WeibullAFT.cs
│   └── LogNormalAFT.cs
├── Ensemble/
│   ├── RandomSurvivalForest.cs
│   └── GradientBoostingSurvival.cs
└── Evaluation/
    ├── ConcordanceIndex.cs
    └── LogRankTest.cs
```

---

### 5. Causal Inference (HIGH PRIORITY - Completely Missing)

**New Interfaces Required:**
```csharp
// src/Interfaces/ICausalModel.cs
public interface ICausalModel<T> : IFullModel<T, Matrix<T>, Vector<T>>
{
    T EstimateAverageTreatmentEffect(Matrix<T> X, Vector<T> treatment);
    Vector<T> EstimateIndividualTreatmentEffects(Matrix<T> X, Vector<T> treatment);
    (T ate, T standardError, T confidenceIntervalLower, T confidenceIntervalUpper)
        EstimateATEWithConfidence(Matrix<T> X, Vector<T> treatment, double confidenceLevel = 0.95);
}

// src/Interfaces/IPropensityScoreModel.cs
public interface IPropensityScoreModel<T> : IFullModel<T, Matrix<T>, Vector<T>>
{
    Vector<T> EstimatePropensityScores(Matrix<T> X);
    Vector<T> GetPropensityWeights(Matrix<T> X, Vector<T> treatment);
}
```

**Directory Structure (NEW):**
```
src/CausalInference/
├── CausalModelBase.cs
├── PropensityScore/
│   ├── PropensityScoreEstimator.cs
│   ├── PropensityScoreMatching.cs
│   └── InversePropensityWeighting.cs
├── Estimators/
│   ├── DoublyRobustEstimator.cs
│   └── AugmentedIPW.cs
├── MetaLearners/
│   ├── SLearner.cs
│   ├── TLearner.cs
│   └── XLearner.cs
└── Forest/
    └── CausalForest.cs
```

---

### 6. Ordinal Regression (MEDIUM PRIORITY)

**New Interface Required:**
```csharp
// src/Interfaces/IOrdinalClassifier.cs
public interface IOrdinalClassifier<T> : IClassifier<T>
{
    Vector<T> Thresholds { get; }
    Vector<T> PredictCumulativeProbabilities(Matrix<T> input);
    int NumOrdinalClasses { get; }
}
```

**Files to Create:**
```
src/Classification/Ordinal/
├── OrdinalClassifierBase.cs  (extends ClassifierBase, implements IOrdinalClassifier)
├── OrdinalLogisticRegression.cs
├── OrdinalRidge.cs
└── AllThresholdOrdinal.cs
```

---

### 7. Online/Incremental Learning (MEDIUM PRIORITY)

**New Interfaces Required:**
```csharp
// src/Interfaces/IOnlineClassifier.cs
public interface IOnlineClassifier<T> : IClassifier<T>
{
    void PartialFit(Matrix<T> X, Vector<T> y);
    void PartialFitSingle(Vector<T> x, T y);
}

// src/Interfaces/IOnlineRegressor.cs
public interface IOnlineRegressor<T> : IRegression<T>
{
    void PartialFit(Matrix<T> X, Vector<T> y);
    void PartialFitSingle(Vector<T> x, T y);
}

// src/Interfaces/IDriftDetector.cs
// Note: Drift detectors are utilities, not prediction models
public interface IDriftDetector<T>
{
    bool DetectDrift(T newValue);
    void Reset();
    bool IsInDrift { get; }
    double DriftProbability { get; }
}
```

**Files to Create:**
```
src/Classification/Online/
├── OnlineClassifierBase.cs  (extends ClassifierBase, implements IOnlineClassifier)
├── HoeffdingTreeClassifier.cs
├── AdaptiveRandomForestClassifier.cs
└── OnlineNaiveBayes.cs

src/DriftDetection/
├── DriftDetectorBase.cs
├── ADWIN.cs
├── DDM.cs
├── EDDM.cs
└── PageHinkley.cs
```

---

### 8. HistGradientBoosting (MEDIUM PRIORITY)

**No new interface needed** - extends existing classifier/regression interfaces.

**Files to Create:**
```
src/Classification/Ensemble/
├── HistGradientBoostingClassifier.cs  (extends EnsembleClassifierBase)

src/Regression/
├── HistGradientBoostingRegressor.cs  (extends RegressionBase)
```

---

### 9. GLMs - Remaining (MEDIUM PRIORITY)

**No new interface needed** - extends existing IRegression.

**Files to Create:**
```
src/Regression/
├── GammaRegressor.cs  (extends RegressionBase)
├── TweedieRegressor.cs  (extends RegressionBase)
```

---

### 10. Time Series Classification (MEDIUM PRIORITY)

**New Interface Required:**
```csharp
// src/Interfaces/ITimeSeriesClassifier.cs
public interface ITimeSeriesClassifier<T> : IClassifier<T>
{
    int SequenceLength { get; }
    int NumChannels { get; }
    void TrainOnSequences(Tensor<T> sequences, Vector<T> labels);
    Vector<T> PredictSequences(Tensor<T> sequences);
}
```

**Files to Create:**
```
src/Classification/TimeSeries/
├── TimeSeriesClassifierBase.cs  (extends ClassifierBase, implements ITimeSeriesClassifier)
├── RocketTransform.cs
├── MiniRocketTransform.cs
├── RocketClassifier.cs
└── TimeSeriesForestClassifier.cs
```

---

### Phase 1: Semi-Supervised & Multi-Label (21 story points)
| Task | Points | Interface |
|------|--------|-----------|
| ISemiSupervisedClassifier | 1 | Extends IClassifier |
| SelfTrainingClassifier | 5 | |
| LabelPropagation | 5 | |
| LabelSpreading | 3 | |
| BinaryRelevance | 5 | Uses existing IMultiLabelClassifier |
| MLkNN | 2 | |

### Phase 2: Imbalanced Learning (21 story points)
| Task | Points | Interface |
|------|--------|-----------|
| ADASYN | 5 | None (augmenter) |
| BorderlineSMOTE | 5 | None (augmenter) |
| SVMSMOTE | 3 | None (augmenter) |
| RandomUnderSampler | 3 | None (augmenter) |
| TomekLinks | 3 | None (augmenter) |
| BalancedRandomForestClassifier | 2 | Extends IClassifier |

### Phase 3: Survival Analysis (21 story points)
| Task | Points | Interface |
|------|--------|-----------|
| ISurvivalModel, ISurvivalEstimator | 2 | Extend IFullModel |
| KaplanMeierEstimator | 5 | |
| CoxProportionalHazards | 8 | |
| NelsonAalenEstimator | 3 | |
| RandomSurvivalForest | 3 | |

### Phase 4: Causal Inference (21 story points)
| Task | Points | Interface |
|------|--------|-----------|
| ICausalModel, IPropensityScoreModel | 2 | Extend IFullModel |
| PropensityScoreEstimator | 5 | |
| InversePropensityWeighting | 3 | |
| DoublyRobustEstimator | 5 | |
| CausalForest | 3 | |
| S/T/X Learners | 3 | |

### Phase 5: Online Learning & Drift Detection (13 story points)
| Task | Points | Interface |
|------|--------|-----------|
| IOnlineClassifier, IOnlineRegressor | 2 | Extend IClassifier/IRegression |
| IDriftDetector | 1 | Standalone utility interface |
| HoeffdingTreeClassifier | 5 | |
| ADWIN drift detector | 3 | |
| DDM drift detector | 2 | |

### Phase 6: HistGradientBoosting & GLMs (13 story points)
| Task | Points | Interface |
|------|--------|-----------|
| HistGradientBoostingClassifier | 5 | Uses IClassifier |
| HistGradientBoostingRegressor | 5 | Uses IRegression |
| GammaRegressor | 1.5 | Uses IRegression |
| TweedieRegressor | 1.5 | Uses IRegression |

### Phase 7: Time Series Classification (8 story points)
| Task | Points | Interface |
|------|--------|-----------|
| ITimeSeriesClassifier | 1 | Extends IClassifier |
| MiniRocketTransform | 3 | |
| RocketClassifier | 4 | |

### Phase 8: Ordinal Regression (5 story points)
| Task | Points | Interface |
|------|--------|-----------|
| IOrdinalClassifier | 1 | Extends IClassifier |
| OrdinalLogisticRegression | 2 | |
| OrdinalRidge | 2 | |

---

## NEW INTERFACES SUMMARY

All new interfaces follow the pattern of extending `IFullModel<T, TInput, TOutput>`:

| Interface | Extends | Purpose |
|-----------|---------|---------|
| `ISemiSupervisedClassifier<T>` | `IClassifier<T>` | Semi-supervised classification |
| `IOrdinalClassifier<T>` | `IClassifier<T>` | Ordinal/ranked classification |
| `IOnlineClassifier<T>` | `IClassifier<T>` | Incremental/streaming classification |
| `IOnlineRegressor<T>` | `IRegression<T>` | Incremental/streaming regression |
| `ISurvivalModel<T>` | `IFullModel<T, Matrix<T>, Vector<T>>` | Survival/time-to-event prediction |
| `ISurvivalEstimator<T>` | `IFullModel<T, (Vector<T>, Vector<T>), Vector<T>>` | Non-parametric survival curves |
| `ICausalModel<T>` | `IFullModel<T, Matrix<T>, Vector<T>>` | Treatment effect estimation |
| `IPropensityScoreModel<T>` | `IFullModel<T, Matrix<T>, Vector<T>>` | Propensity score estimation |
| `ITimeSeriesClassifier<T>` | `IClassifier<T>` | Time series classification |
| `IDriftDetector<T>` | (standalone) | Concept drift detection utility |

**Note:** `IMultiLabelClassifier<T>` already exists and extends `IClassifier<T>`.

---

### 1. Interface Inheritance (CRITICAL)
- [ ] ALL model interfaces MUST extend `IFullModel<T, TInput, TOutput>` or an interface that does
- [ ] Use `IClassifier<T>` for classification-related interfaces
- [ ] Use `IRegression<T>` for regression-related interfaces

### 2. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected INumericOperations<T> NumOps { get; }` initialized via `MathHelper.GetNumericOperations<T>()`
- [ ] NEVER hardcode numeric types

### 3. Base Class Pattern (REQUIRED)
- [ ] Create `{FeatureName}Base.cs` extending appropriate base class
- [ ] Classifiers extend `ClassifierBase<T>`
- [ ] Regressors extend `RegressionBase<T>`

### 4. Random Number Generation (CRITICAL)
- [ ] ALWAYS use `RandomHelper.CreateSeededRandom(seed)` or `RandomHelper.CreateSecureRandom()`

---

## PRIORITY SUMMARY

| Priority | Models Count | Total Story Points |
|----------|--------------|-------------------|
| High | 25 models | ~55 points |
| Medium | 20 models | ~40 points |
| Low | 15 models | ~20 points |
| **Total** | **60 models** | **~115 points** |

---

## Definition of Done

- [ ] All new interfaces extend IFullModel (directly or indirectly)
- [ ] All base classes follow existing patterns (ClassifierBase, RegressionBase)
- [ ] Comprehensive unit tests (80%+ coverage)
- [ ] XML documentation with `<b>For Beginners:</b>` sections
- [ ] All builds pass (net8.0 and net471)
